tests/helper/parameter_checks.hpp

```diff
@@ -0,0 +1,146 @@
+        REQUIRE(element.template get<ParamT>() == expectedAfterSet);
+      }
+    }
+
+    template <typename ParamT, typename ElementT, typename ValueT,
+              typename ComparedT>
+    void check_default_impl(ElementT element, ValueT modifiedVal,
+                            ComparedT expectedDefault,
+                            ComparedT expectedAfterSet) {
+      using boost::typeindex::type_id;
+      INFO("element type: " << type_id<ElementT>().pretty_name());
+      INFO("parameter type: " << type_id<ParamT>().pretty_name());
+      SECTION("check default constructed element") {
+        REQUIRE(element.template has<ParamT>() == true);
+        REQUIRE(element.template isDefault<ParamT>() == true);
+        REQUIRE(element.template get<ParamT>() == expectedDefault);
+      }
+      SECTION("check parameter after set") {
+        element.set(ParamT(modifiedVal));
+        REQUIRE(element.template has<ParamT>() == true);
+        REQUIRE(element.template isDefault<ParamT>() == false);
+        REQUIRE(element.template get<ParamT>() == expectedAfterSet);
+        SECTION("and unset") {
+          element.template unset<ParamT>();
+          REQUIRE(element.template has<ParamT>() == true);
+          REQUIRE(element.template isDefault<ParamT>() == true);
+          REQUIRE(element.template get<ParamT>() == expectedDefault);
+        }
+      }
+    }
+
+    template <typename ParamT, typename ElementT, typename ValueT,
+              typename ComparedT>
+    void check_optional_impl(ElementT element, ValueT modifiedVal,
+                             ComparedT expectedAfterSet) {
+      using boost::typeindex::type_id;
+      INFO("element type: " << type_id<ElementT>().pretty_name());
+      INFO("parameter type: " << type_id<ParamT>().pretty_name());
+      SECTION("check default constructed element") {
+        REQUIRE(element.template has<ParamT>() == false);
+        REQUIRE(element.template isDefault<ParamT>() == false);
+      }
+      SECTION("check parameter after set") {
+        element.set(ParamT{modifiedVal});
+        REQUIRE(element.template has<ParamT>() == true);
+        REQUIRE(element.template isDefault<ParamT>() == false);
+        REQUIRE(element.template get<ParamT>() == expectedAfterSet);
+        SECTION("and unset") {
+          element.template unset<ParamT>();
+          REQUIRE(element.template has<ParamT>() == false);
+          REQUIRE(element.template isDefault<ParamT>() == false);
+        }
+      }
+    }
+    template <typename ValueT,
+              typename std::enable_if<!std::is_floating_point<ValueT>::value,
+                                      bool>::type = true>
+    ValueT wrapFloat(ValueT value) {
+      return value;
+    }
+
+    template <typename ValueT,
+              typename std::enable_if<std::is_floating_point<ValueT>::value,
+                                      bool>::type = true>
+    Approx wrapFloat(ValueT value) {
+      return Approx(value);
```

Comment:
Is this necessary? I know the wisdom is "don't compare floats with ==", but I really think that should be "don't compare floats with ==, unless you know they should be equal". In cases where floats should be passed through without modifications, you can potentially catch bugs that modify things slightly (e.g. unintended conversions).

Reply:
Yeah, good point - I guess we expect values to get passed through as-is and probably want to know if that doesn't happen. The flip-side is that removing the Approx would mean that forgetting to use a float in a test case would fail sometimes (e.g. if you chose 0.1), but not others (e.g. 1.0). I think we can partially get around that by making the parameter construction use braced initialisation (I had this initially but relaxed it). Unfortunately only MSVC seems to catch that the narrowing from double->float in that context as a compile time error. At least you'd find out when you hit CI I guess?

Reply:
If it can be made to not accept the wrong type that would be nicer IMO.

> Unfortunately only MSVC seems to catch that the narrowing from double->float in that context as a compile time error.

I think that can be enabled with -Wconversion; might be worth turning that on by default as it's annoying finding out about those warnings in CI.

It's worth remembering too that these parameters should really all be doubles, which would also get rid of this issue, I think.

Reply:
Ah that's a useful flag. I'm a bit iffy about setting warning flags by default as it can be a pain when you consume someone else's project. I guess we could only do it when the project is at the root level? 
I wonder if this is a use case for the new CMake presets feature, as could have a 'dev' preset with the warnings cranked up and a default one that's more relaxed.